{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: minizinc in c:\\users\\32057\\anaconda3\\lib\\site-packages (0.9.0)Note: you may need to restart the kernel to use updated packages.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pip install minizinc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "the synchronous MiniZinc Python `solve()` method was called from an asynchronous environment.\n\nBecause Python's asyncio library does not support using multiple event loops that would be required to use this method, we instead suggest you use the `solve_async()' method or patch Python behaviour with a package such as `nested_asyncio'.\n\nOriginal message: asyncio.run() cannot be called from a running event loop",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\minizinc\\instance.py\u001b[0m in \u001b[0;36msolve\u001b[1;34m(self, timeout, nr_solutions, processes, random_seed, all_solutions, intermediate_solutions, free_search, optimisation_level, **kwargs)\u001b[0m\n\u001b[0;32m    182\u001b[0m                 \u001b[0masyncio\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mset_event_loop_policy\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0masyncio\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mWindowsProactorEventLoopPolicy\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 183\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0masyncio\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcoroutine\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    184\u001b[0m         \u001b[1;32mexcept\u001b[0m \u001b[0mRuntimeError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\asyncio\\runners.py\u001b[0m in \u001b[0;36mrun\u001b[1;34m(main, debug)\u001b[0m\n\u001b[0;32m     32\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mevents\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_running_loop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 33\u001b[1;33m         raise RuntimeError(\n\u001b[0m\u001b[0;32m     34\u001b[0m             \"asyncio.run() cannot be called from a running event loop\")\n",
      "\u001b[1;31mRuntimeError\u001b[0m: asyncio.run() cannot be called from a running event loop",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-2d742a676e85>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m    135\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    136\u001b[0m \u001b[1;31m#nest_asyncio.apply()\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 137\u001b[1;33m \u001b[0mres\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0msolve_mcp\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"./CP.mzn\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\"../instances_dnz/inst01.dnz\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    138\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mres\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-1-2d742a676e85>\u001b[0m in \u001b[0;36msolve_mcp\u001b[1;34m(custom_model, file_path)\u001b[0m\n\u001b[0;32m    130\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    131\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 132\u001b[1;33m   \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0minstance\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msolve\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    133\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    134\u001b[0m   \u001b[1;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\minizinc\\instance.py\u001b[0m in \u001b[0;36msolve\u001b[1;34m(self, timeout, nr_solutions, processes, random_seed, all_solutions, intermediate_solutions, free_search, optimisation_level, **kwargs)\u001b[0m\n\u001b[0;32m    184\u001b[0m         \u001b[1;32mexcept\u001b[0m \u001b[0mRuntimeError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    185\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[1;34m\"called from a running event loop\"\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 186\u001b[1;33m                 raise RuntimeError(\n\u001b[0m\u001b[0;32m    187\u001b[0m                     \u001b[1;34m\"the synchronous MiniZinc Python `solve()` method was called from\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    188\u001b[0m                     \u001b[1;34m\" an asynchronous environment.\\n\\nBecause Python's asyncio library\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mRuntimeError\u001b[0m: the synchronous MiniZinc Python `solve()` method was called from an asynchronous environment.\n\nBecause Python's asyncio library does not support using multiple event loops that would be required to use this method, we instead suggest you use the `solve_async()' method or patch Python behaviour with a package such as `nested_asyncio'.\n\nOriginal message: asyncio.run() cannot be called from a running event loop"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Sat Aug 17 22:39:52 2024\n",
    "\n",
    "@author: 32057\n",
    "\"\"\"\n",
    "from minizinc import Instance, Model, Solver\n",
    "import minizinc\n",
    "import re\n",
    "\n",
    "#import nest_asyncio\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Function to parse the value after the equal sign\n",
    "def parse_value(value):\n",
    "    return value.strip().strip(';')\n",
    "    \n",
    "\"\"\"\n",
    "def read_instance1(file_path):\n",
    "    # Initialize data structures\n",
    "    m = None\n",
    "    n = None\n",
    "    l = []\n",
    "    s = []\n",
    "    distances = []\n",
    "    data = []\n",
    "    # Read the data file\n",
    "    with open(file_path, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "        for line in lines:\n",
    "            line = line.strip()\n",
    "            data.append(line)\n",
    "            if line.startswith('m'):\n",
    "                m = int(parse_value(line.split('=')[1]))\n",
    "            elif line.startswith('n'):\n",
    "                n = int(parse_value(line.split('=')[1]))\n",
    "            elif line.startswith('l'):\n",
    "                l = parse_value(line.split('=')[1])\n",
    "            elif line.startswith('s'):\n",
    "                s = parse_value(line.split('=')[1])\n",
    "            elif line.startswith('D') or line.startswith('|'):\n",
    "                # Start reading the distances array\n",
    "\n",
    "                #distance_lines = distance_lines.strip(\" \")[1:]\n",
    "                distance_lines = line.split('=')[1]\n",
    "                row = []\n",
    "                for c in distance_lines:\n",
    "                    if c.isnumeric():\n",
    "                        row.append(int(c))\n",
    "\n",
    "                distances.append(row)\n",
    "                for line in lines:\n",
    "                    print(\"line\")\n",
    "                    \n",
    "                    #row = parse_value(line.split('=')[1])\n",
    "                    #print(f\"row:{row}{type(row)}\")\n",
    "                \n",
    "                for i in range(n+1):\n",
    "                    if not distance_lines:\n",
    "                        distance_lines = next(file).strip()\n",
    "                    row = parse_value(distance_lines.split('],')[0] + ']')\n",
    "                    distances.append(row)\n",
    "                    distance_lines = next(file).strip()\n",
    "    \n",
    "    # Display parsed data\n",
    "    print(\"m =\", m)\n",
    "    print(\"n =\", n)\n",
    "    print(\"l =\", l)\n",
    "    print(\"s =\", s)\n",
    "    print(\"distances =\", distances)\n",
    "    return m, n, l, s, distances\n",
    "\"\"\"\n",
    "def read_instance(file_path):\n",
    "    # Initialize data structures\n",
    "    m = None\n",
    "    n = None\n",
    "    l = []\n",
    "    s = []\n",
    "    distances = []\n",
    "    data = []\n",
    "    # Read the data file\n",
    "    with open(file_path, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "        m = int(parse_value(lines[0].split('=')[1]))\n",
    "        n = int(parse_value(lines[1].split('=')[1]))\n",
    "        \n",
    "        for c in re.findall(r'\\b\\d+\\b', parse_value(lines[2].split('=')[1])):\n",
    "            l.append(int(c))\n",
    "        \n",
    "        for c in re.findall(r'\\b\\d+\\b', parse_value(lines[3].split('=')[1])):\n",
    "            s.append(int(c))\n",
    "    \n",
    "        for line in lines[4:]:\n",
    "            row = []\n",
    "            #print(line.split(','))\n",
    "            for el in line.split(','):\n",
    "                r =re.findall(r'\\b\\d+\\b', el)\n",
    "                for c in r: \n",
    "                    row.append(int(c))\n",
    "            distances.append(row)\n",
    "    return m, n, l, s, distances\n",
    "\n",
    "def solve_mcp(custom_model, file_path):\n",
    "  #m, n, l, s, D = read_instance(file_path)\n",
    "  \n",
    "  m = \"3;\"\n",
    "  n = \"7;\"\n",
    "  l = \"[15, 10, 7];\"\n",
    "  s = \"[3, 2, 6, 8, 5, 4, 4];\"\n",
    "  D =\"[|0, 3, 3, 6, 5, 6, 6, 2 | 3, 0, 4, 3, 4, 7, 7, 3 | 3, 4, 0, 7, 6, 3, 5, 3 | 6, 3, 7, 0, 3, 6, 6, 4 | 5, 4, 6, 3, 0, 3, 3, 3 | 6, 7, 3, 6, 3, 0, 2, 4 | 6, 7, 5, 6, 3, 2, 0, 4 | 2, 3, 3, 4, 3, 4, 4, 0 |];\"\n",
    "  LB = \"0;\"\n",
    "  UB = \"50;\"\n",
    "  # Load model\n",
    "  model = minizinc.Model(custom_model)\n",
    "\n",
    "  gecode = minizinc.Solver.lookup(\"gecode\")\n",
    "  # Create minizinc instance\n",
    "  instance = minizinc.Instance(gecode, model)\n",
    "  instance[\"m\"] = m\n",
    "  instance[\"n\"] = n\n",
    "  instance[\"l\"] = l\n",
    "  instance[\"s\"] = s\n",
    "  instance[\"D\"] = D\n",
    "  instance[\"LB\"] = LB\n",
    "  instance[\"UB\"] = UB\n",
    "  #instance[\"o\"] = origin_location\n",
    "  # Solve the problem\n",
    "  \n",
    "\n",
    "  result = instance.solve()\n",
    "\n",
    "  return result\n",
    "\n",
    "#nest_asyncio.apply()\n",
    "res = solve_mcp(\"./CP.mzn\",\"../instances_dnz/inst01.dnz\")\n",
    "print(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
